Testing workflow for 4.x

Bouncheck opened the pull request:
Translated 3.x workflow file for 4.x driver version.
Looks roughly the same except few minor changes.

Bouncheck committed:
Testing workflow for 4.x

Translated 3.x workflow file for 4.x driver version.
Looks roughly the same except few minor changes.

avelanarius commented:
You are probably aware of it, but Scylla Enterprise test fail due to: https://github.com/scylladb/java-driver/commit/0c262f8f05eb59f68fcd8dc5f0ca44118447068a not being ported. `5.0.rc7` probably fails due to: https://github.com/scylladb/java-driver/commit/17e7374a5633771dd2512b7e7b444b5ef172811f not being ported.

I would also appreciate testing Cassandra 4 (`version_fetch.py` should be extended with it - it's the same as Cassandra 3 except for regex (probably?):
```
CASSANDRA3_REGEX = re.compile(r'a href="(3)\.(\d+)\.(\d+)/"')
CASSANDRA4_REGEX = re.compile(r'a href="(4)\.(\d+)\.(\d+)/"')
```

Bouncheck committed:
Port 0c262f8 from 3.x version branch.

Adds Environment Map to CcmBride. Now it will use it during .execute() calls.

Bouncheck committed:
Port 17e7374 from 3.x branch (Parse rc versions)

Bouncheck committed:
Add Cassandra4 version fetch and test against it.

avelanarius commented:
I saw that tests on 2020.1.13 failed on LWT in two separate runs - I think that the reason is that those tests rely on `SCYLLA_LWT_ADD_METADATA_MARK` and this extension is not implemented in 2020.1.13 (see commit https://github.com/scylladb/scylla/commit/6028588148c556151d764f2a987c276cb4a21eb0 in enterprise repo - it's missing on 2020.*). 

Two solutions:
1. Introduce something similar to: https://github.com/scylladb/java-driver/blob/scylla-4.x/test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/CassandraRequirement.java - `ScyllaRequirement` so that the test is skipped on 2020.* (make sure that the requirement has both minimum OSS and Enterprise version).
2. Reduce the number of Scylla Enterprise versions we test - 2020 is old and "solution 1" can be done in the future as a follow-up.

Bouncheck committed:
Add ScyllaRequirement annotation and apply it.

Adds new annotation. Allows for both OSS and Enterprise version requirements.
Annotates some tests with it.

Bouncheck committed:
Pass differently parsed Scylla version during ccm creation.

Version.parse automatically fills patch number as 0 if not present.
This leads to some problems with ccm when setup with (e.g.) 5.0-rcX version.
This change makes method returning version string remove patch number back
when dealing with such version number.

Bouncheck committed:
Mention new annotation in CONTRIBUTING.md

Bouncheck committed:
Mark and skip flaky test

avelanarius commented:
Merged outside GitHub to `scylla-4.x` branch. Note that I observed that there are still some flaky tests (to be disabled)...
